### ISSUE TITLE
Fix AuthorizationConfigControllerTestIT.roleShouldBeDeleted

### ISSUE THREAD

**Issue by sbespalov** [strongbox/strongbox#1744]
# Task Description

We need to fix this test against the `issues/1649` branch.

Failed Jenkins build [here](https://jenkins.carlspring.org/blue/organizations/jenkins/strongbox%2Fbuilds%2Fstrongbox/detail/PR-1676/108/tests)

The problem seems related with JG issue that returrns single object value for `Cardinality.set` properties that contains only one element in collection. As a workaround we currently always add empty string into property collection to have more then one element in collection. But with current implementation its still posible to have single empty string element in collection. This should be fixed, and if original entity collection was empty then there is no need to add emty sting into property collection, so property collection should be empty as well.
Also we should improve this workaround with custom `EntityTraversalDsl` method (see below tasks).

# Tasks

The following tasks will need to be carried out:
* [ ] create `<E> EntityTraversalDsl#property(String key, Set<E> value)` method
* [ ] use `EntityTraversalDsl#property(String key, Set<E> value)` method within `VertexEntityTraversalAdapter` classes instead of common `property` method with `Cardinality.set`

# Task Relationships

This task:
* Is a sub-task of: #1649

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @steve-todorov

**Comment by anki2189** [comment]
@sbespalov @carlspring I will pick this.

**Comment by carlspring** [comment]
I see that @sbespalov has marked this as `on hold`. Pehaps you could pick one of the other ones?

@sbespalov : What is this blocked by?

**Comment by sbespalov** [comment]
@anki2189 I've updated description, plees proceed with it if you like.

**Comment by anki2189** [comment]
Sure @sbespalov

**Comment by anki2189** [comment]
@sbespalov Issue seems different here.
Exception `java.lang.ClassCastException: java.lang.String cannot be cast to java.util.Collection` is due to conversion of `CUSTOM_ROLE` (role queried for using IN clause) to collection.
Not sure this is related to https://stackoverflow.com/questions/49663376/cypher-query-for-searching-through-a-list or not but my assumption is we need to fix the way we are quering for Set Cardinality property.

**Comment by sbespalov** [comment]
> @sbespalov Issue seems different here.
> Exception `java.lang.ClassCastException: java.lang.String cannot be cast to java.util.Collection` is due to conversion of `CUSTOM_ROLE` (role queried for using IN clause) to collection.

@anki2189 can you please clarify your guess? 

The cause described in issue description and corresponds to the error log in Jenkins.
I would suggest to proceed with proposed solution and see how it works.

**Comment by anki2189** [comment]
> > @sbespalov Issue seems different here.
> > Exception `java.lang.ClassCastException: java.lang.String cannot be cast to java.util.Collection` is due to conversion of `CUSTOM_ROLE` (role queried for using IN clause) to collection.
> 
> @anki2189 can you please clarify your guess?
> 
Attaching debugger screenshots where this is failing
<img width="1170" alt="Screen Shot 2020-04-20 at 10 21 45 AM" src="https://user-images.githubusercontent.com/7271803/79715408-3e2bab80-82f1-11ea-8fbb-b49586f82b23.png">
<img width="1169" alt="Screen Shot 2020-04-20 at 10 23 08 AM" src="https://user-images.githubusercontent.com/7271803/79715409-3ff56f00-82f1-11ea-8c8f-fe881da57443.png">

**Comment by sbespalov** [comment]
@anki2189 the problem is not with `CUSTOM_ROLE` value, if you'll check what huppens inside `predicate.test` method then you will see that it tryes to cast an empty string to collection.

**Comment by anki2189** [comment]
@sbespalov Tried without empty string and two elements in `user.roles` but still getting Same Exception.

**Comment by sbespalov** [comment]
could you please create `UserRepositoryTest` with cases for `findUsersWithRole`?

**Comment by anki2189** [comment]
@sbespalov created `UserRepositoryTest`

**Comment by sbespalov** [comment]
the problem may be related with `cypher-for-gremlin`  bug, so now the resulting query is:
```
Gremlin: g.V().as('user').hasLabel('User').where(__.select('user').values('roles').as('  GENERATED1').constant('ADMIN').where(within('  GENERATED1'))).select('user').project('user').by(__.valueMap().with('~tinkerpop.valueMap.tokens'))
```
but most likely it should be:
```
Gremlin: g.V().as('user').hasLabel('User').where(__.select('user').values('roles').unfold().as('  GENERATED1').constant('ADMIN').where(within('  GENERATED1'))).select('user').project('user').by(__.valueMap().with('~tinkerpop.valueMap.tokens'))
```
I propose to do as follows:
1. The `EntityTraversalDsl` part should be done within this task as it's seems needed anyway.
2. I'll create a followup task to rework `User.roles` into standalone Vertex related with `User`, this seems required as well to search user by role using index.
 
@anki2189 would you like to proceed with `EntityTraversalDsl` here?

**Comment by anki2189** [comment]
@sbespalov Changes done as suggested, pls review.

**Comment by carlspring** [comment]
Thanks for your help, @anki2189 , @sbespalov ! :)
